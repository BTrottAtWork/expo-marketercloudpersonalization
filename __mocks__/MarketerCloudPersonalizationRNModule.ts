import { LocalEventEmitter } from '../src/utils/localEventEmitter';

let mockEventEmitter = new LocalEventEmitter();

const mockNativeModule = {
    listenForCampaignTarget: jest.fn(),
    stopListeningForCampaignTarget: jest.fn(),
    getAccountId: jest.fn(),
    setAccountId: jest.fn(),
    getAnonymousId: jest.fn(),
    getUserId: jest.fn(),
    setUserId: jest.fn(),
    viewItem: jest.fn(),
    viewCategory: jest.fn(),
    trackAction: jest.fn(),
    addToCart: jest.fn(),
    trackClickthrough: jest.fn(),
    trackDismissal: jest.fn(),
    trackImpression: jest.fn(),
    removeListener: jest.fn(),
    addListener: jest.fn((eventName, callback) => {
        const unsub = mockEventEmitter.addListener(eventName, callback);

        return {
            remove: () => {
                unsub();
            }
        }
    }),
    removeListeners: jest.fn(),
};

export default mockNativeModule;

export const emitMockEvent = (eventName: string, data: any) => {
    mockEventEmitter.emit(eventName, data);
};
