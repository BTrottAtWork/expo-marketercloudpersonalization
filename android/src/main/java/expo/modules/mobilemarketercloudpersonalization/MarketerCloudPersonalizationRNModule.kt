package expo.modules.mobilemarketercloudpersonalization

import com.evergage.android.Campaign
import com.evergage.android.Evergage
import com.evergage.android.promote.Category
import com.evergage.android.promote.LineItem
import com.evergage.android.promote.Product
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.ConcurrentHashMap
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

class MarketerCloudPersonalizationRNModule : Module() {

  private var listeners: MutableSet<String> = mutableSetOf();
  private var activeCampaigns: MutableMap<String, Campaign> = ConcurrentHashMap<String, Campaign>();

  override fun definition() = ModuleDefinition {
    Name("MarketerCloudPersonalizationRN")
    Events("mcp_data_campaign")

    AsyncFunction("stopListeningForCampaignTarget") { campaignTarget: String ->
      val evergage = Evergage.getInstance()
      val context = evergage.globalContext
      context?.setCampaignHandler(null, campaignTarget)

      if(listeners.contains(campaignTarget)) {
        listeners.remove(campaignTarget)
      }
    }

    AsyncFunction("listenForCampaignTarget") { campaignTarget: String ->
      val evergage = Evergage.getInstance()
      val context = evergage.globalContext

      val handler = { campaign: Campaign ->
          activeCampaigns[campaign.campaignId] = campaign;

          val gson = Gson()
          val type = object : TypeToken<Map<String, Any>>() {}.type
          val data: Map<String, Any> = gson.fromJson(campaign.data.toString(), type)

          sendEvent("mcp_data_campaign", mapOf(
              "target" to campaignTarget,
              "campaignId" to campaign.campaignId,
              "campaignName" to campaign.campaignName,
              "experienceId" to campaign.experienceId,
              "experienceName" to campaign.experienceName,
              "messageId" to campaign.messageId,
              "isControlGroup" to campaign.isControlGroup,
              "data" to data
          ))
      }

      context?.setCampaignHandler(handler, campaignTarget)
      listeners.add(campaignTarget)
    }

    AsyncFunction("getAccountId") {
        return@AsyncFunction Evergage.getInstance().accountId
    }

    AsyncFunction("setAccountId") { value: String ->
      Evergage.getInstance().accountId = value
    }

    AsyncFunction("getAnonymousId") {
        return@AsyncFunction Evergage.getInstance().anonymousId
    }

    AsyncFunction("getUserId") {
        return@AsyncFunction Evergage.getInstance().userId
    }

    Function("setUserId") { value: String ->
      Evergage.getInstance().userId = value
    }

    AsyncFunction("viewItem") { id: String ->
     Evergage.getInstance().globalContext?.viewItem(Product(id));
    }

    AsyncFunction("viewCategory") { id: String ->
      Evergage.getInstance().globalContext?.viewCategory(Category(id))
    }

    AsyncFunction("trackAction") { action: String ->
     Evergage.getInstance().globalContext?.trackAction(action)
    }

    AsyncFunction("addToCart") { productId: String, quantity: Int ->
      Evergage.getInstance().globalContext?.addToCart(LineItem(Product(productId), quantity))
    }

    AsyncFunction("trackClickthrough") { campaignId: String ->
        activeCampaigns[campaignId]?.let {
          Evergage.getInstance().globalContext?.trackClickthrough(it)
        }
    }

    AsyncFunction("trackDismissal") { campaignId: String ->
      activeCampaigns[campaignId]?.let {
        Evergage.getInstance().globalContext?.trackDismissal(it);
      }
    }

    AsyncFunction("trackImpression") { campaignId: String ->
      activeCampaigns[campaignId]?.let {
        Evergage.getInstance().globalContext?.trackImpression(it);
      }
    }

  }
}
