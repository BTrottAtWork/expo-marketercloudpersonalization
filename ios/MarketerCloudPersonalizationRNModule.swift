import ExpoModulesCore
import Evergage

public class MarketerCloudPersonalizationRNModule: Module {
    
  private var activeCampaigns: [String: EVGCampaign] = [:]
  private var listeners: Set<String> = []
    
  private func campaignHandler (campaign: EVGCampaign) {
      
      NSLog("MarketerCloudPersonalizationRNModule: campaignHandler: \(campaign)")
      
      if listeners.contains(campaign.target) {
          activeCampaigns[campaign.campaignId] = campaign
          
          let payload: [String: Any] = [
            "target": campaign.target,
            "campaignId": campaign.campaignId,
            "campaignName": campaign.campaignName,
            "experienceId": campaign.experienceId,
            "experienceName": campaign.experienceName,
            "messageId": campaign.messageId,
            "isControlGroup": campaign.isControlGroup,
            "data": campaign.data
          ]
          
          self.sendEvent("mcp_data_campaign", payload)
      }
  }
    
  public func definition() -> ModuleDefinition {
      Name("MarketerCloudPersonalizationRN")

      Events("mcp_data_campaign")
                  
      AsyncFunction("stopListeningForCampaignTarget") { (campaignTarget: String) in
          let evergage = Evergage.sharedInstance()
          if let context = evergage.globalContext {
              context.setCampaignHandler(nil, forTarget: campaignTarget)
          }
          
          if self.listeners.contains(campaignTarget) {
              self.listeners.remove(campaignTarget)
          }
      }
      
      AsyncFunction("listenForCampaignTarget") { (campaignTarget: String) in
        let evergage = Evergage.sharedInstance()
        weak var weakSelf = self
        if let context = evergage.globalContext {
          context.setCampaignHandler(weakSelf?.campaignHandler, forTarget: campaignTarget)
        }

        self.listeners.insert(campaignTarget)
      }
      
      AsyncFunction("getAccountId") {
        let evergage = Evergage.sharedInstance()
        return evergage.accountId
      }
    
      AsyncFunction("setAccountId") { (value: String) in
        let evergage = Evergage.sharedInstance()
        evergage.accountId = value
      }
                    
      AsyncFunction("getAnonymousId") {
        let evergage = Evergage.sharedInstance()
        return evergage.anonymousId
      }
      
      AsyncFunction("getUserId") {
        let evergage = Evergage.sharedInstance()
        return evergage.userId
      }
                    
      Function("setUserId") { (value: String) in
        let evergage = Evergage.sharedInstance()
        evergage.userId = value
      }
      
      AsyncFunction("viewItem") { (id: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            context.viewItem(EVGProduct(id: id))
        }
      }
      
      AsyncFunction("viewCategory") { (id: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            context.viewCategory(EVGCategory(id: id))
        }
      }
      
      AsyncFunction("trackAction") { (action: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
          context.trackAction(action)
        }
      }
      
      AsyncFunction("addToCart") { (productId: String, quantity: Int) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            context.add(toCart: EVGLineItem(item: EVGProduct(id: productId), quantity: NSNumber(value: quantity)))
        }
      }
      
      AsyncFunction("trackClickthrough") { (campaignId: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            if let c = activeCampaigns[campaignId] {
                context.trackClickthrough(c)
            }
        }
      }
      
      AsyncFunction("trackDismissal") { (campaignId: String) in
          let evergage = Evergage.sharedInstance()
          if let context = evergage.globalContext {
              if let c = activeCampaigns[campaignId] {
                  context.trackDismissal(c)
              }
          }
      }
      
      AsyncFunction("trackImpression") { (campaignId: String) in
          let evergage = Evergage.sharedInstance()
          if let context = evergage.globalContext {
              if let c = activeCampaigns[campaignId] {
                  context.trackImpression(c)
              }
          }
      }
  }
}
