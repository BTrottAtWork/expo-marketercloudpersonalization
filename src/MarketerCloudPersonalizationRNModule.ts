import { requireNativeModule, NativeModule } from "expo-modules-core";
import { MarketerCloudPersonalizationRNModuleEvents } from "./MarketerCloudPersonalizationRN.types";

declare class MarketerCloudPersonalizationRNModule extends NativeModule<
  MarketerCloudPersonalizationRNModuleEvents<any>
> {
  getAccountId(): Promise<string>;
  setAccountId(accountId: string): Promise<void>;
  getAnonymousId(): Promise<string>;
  getUserId(): Promise<string>;
  setUserId(userId: string): void;
  viewItem(itemId: string): Promise<void>;
  viewCategory(categoryId: string): Promise<void>;
  trackAction(action: string): Promise<void>;
  addToCart(productId: string, quantity: number): Promise<void>;
  trackClickthrough(campaignId: string): Promise<void>;
  trackDismissal(campaignId: string): Promise<void>;
  trackImpression(campaignId: string): Promise<void>;
  listenForCampaignTarget(campaignTarget: string): Promise<void>;
  stopListeningForCampaignTarget(campaignTarget: string): Promise<void>;
}

// This call loads the native module object from the JSI.
export default requireNativeModule<MarketerCloudPersonalizationRNModule>(
  "MarketerCloudPersonalizationRN",
);
