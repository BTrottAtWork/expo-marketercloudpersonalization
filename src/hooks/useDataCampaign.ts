import { EventSubscription } from "expo-modules-core";
import { useState, useEffect, useRef } from "react";

import { addListenerForCampaignTarget } from "..";
import {
  DataCampaignEvent,
  DataCampaignPrioritizationOptions,
} from "../MarketerCloudPersonalizationRN.types";
import { defaultPrioritizationHandler } from "../utils/dataCampaignPrioritization";

/**
 *
 * This hook establishes a listener for 1:M campaign targets, with MCP
 *
 * As we notify MCP of app events (actions, views, dismissals, etc...) MCP will respond with
 * campaign events where conditions are met for this user.
 *
 * Campaigns are stored and updated via the monitoring hook.
 *
 * @param targets
 * @returns
 */
export const useDataCampaign = <DataCampaignPayloadType>(
  targets: string[],
  prioritizationOptions?: DataCampaignPrioritizationOptions,
  customPrioritizationHandler?: (
    campaigns: DataCampaignEvent<DataCampaignPayloadType>[],
    opts?: DataCampaignPrioritizationOptions,
  ) => DataCampaignEvent<DataCampaignPayloadType>[],
) => {
  const data = useRef({});
  const [campaigns, setCampaigns] = useState<
    DataCampaignEvent<DataCampaignPayloadType>[]
  >([]);
  const [campaignReady, setCampaignReady] = useState(false);
  const listenersRef = useRef<Record<string, EventSubscription>>({});

  const removeCampaign = (campaignId: string): void => {
    const camps = campaigns.filter((c) => c.campaignId !== campaignId);
    setCampaigns(camps);
    setCampaignReady(camps.length > 0);
  };

  const clearCampaigns = () => {
    setCampaigns([]);
    data.current = {};
    setCampaignReady(false);
  };

  useEffect(() => {
    targets.forEach((t) => {
      listenersRef.current[t] = addListenerForCampaignTarget(
        t,
        (
          campaignData: DataCampaignEvent<DataCampaignPayloadType | string>,
        ): void => {
          const updatedData = { ...data.current };
          updatedData[t] = campaignData;
          data.current = updatedData;

          const activeTargets = targets.filter((t) => data.current[t]);
          const activeCampaigns = activeTargets.map((t) => data.current[t]);
          const prioritizedCampaigns = customPrioritizationHandler
            ? customPrioritizationHandler(
                activeCampaigns,
                prioritizationOptions,
              )
            : defaultPrioritizationHandler(
                activeCampaigns,
                prioritizationOptions,
              );

          setCampaigns(prioritizedCampaigns);
          setCampaignReady(prioritizedCampaigns.length > 0);
        },
      );
    });

    return () => {
      Object.values(listenersRef.current).forEach(
        (l: EventSubscription): void => l.remove(),
      );
      clearCampaigns();
    };
  }, []);

  return { ready: campaignReady, campaigns, removeCampaign, clearCampaigns };
};
