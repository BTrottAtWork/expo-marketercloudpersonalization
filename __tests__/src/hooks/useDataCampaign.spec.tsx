import { useDataCampaign } from "../../../src/hooks/useDataCampaign";
import {  act, renderHook } from "@testing-library/react-native";
import { emitMockEvent } from "../../../__mocks__/MarketerCloudPersonalizationRNModule";

const mockCampaignEvent = {
    target: "my_target", 
    campaignId: "my_id", 
    data: {
        text: "hello world"
    }       
}

export interface TestCampaignType {
    text: string;
};

describe("useDataCampaign", () => {

    test("No campaigns fired should render null", () => {
        const {result} = renderHook(() => useDataCampaign<TestCampaignType>(["my_target", "remove_one", "remove_all"]));
        expect(result.current.ready).toBeFalsy();
        expect(result.current.campaigns).toMatchSnapshot();
    });

    test("Campaign fired for another target should render null", () => {
        const {result} = renderHook(() => useDataCampaign<TestCampaignType>(["my_target", "remove_one", "remove_all"]));
        act(() => {
            emitMockEvent("mcp_data_campaign", {...mockCampaignEvent, target: "not_my_target"});
        })
    
        expect(result.current.ready).toBeFalsy();
        expect(result.current.campaigns).toMatchSnapshot();
    });

    test("Campaign fired for my target should render hello world", () => {
        const {result} = renderHook(() => useDataCampaign<TestCampaignType>(["my_target", "remove_one", "remove_all"]));
        expect(result.current.ready).toBeFalsy();
        act(()=> {
            emitMockEvent("mcp_data_campaign", mockCampaignEvent);
        })
        expect(result.current.ready).toBeTruthy();
        expect(result.current.campaigns.length).toBeGreaterThan(0);
    });

    test("continue to render hello world when single campaign is removed.", () => {
        const {result} = renderHook(() => useDataCampaign<TestCampaignType>(["my_target", "dont_care", "remove"]));
        
        act(() => {
            emitMockEvent("mcp_data_campaign", mockCampaignEvent);
            emitMockEvent("mcp_data_campaign", {...mockCampaignEvent, target: "data_event_2", campaignId: "two"});
            emitMockEvent("mcp_data_campaign", {...mockCampaignEvent, target: "remove", campaignId: "three"}); 
        })

        act(() => {
            result.current.removeCampaign("three");
        })
        
        expect(result.current.ready).toBeTruthy();
        expect(result.current.campaigns).toMatchSnapshot();
    });

    test("render null when all campaigns are removed.", () => {
        const {result} = renderHook(() => useDataCampaign<TestCampaignType>(["my_target", "remove_one", "remove_all"]));
        
        act(() => {
            emitMockEvent("mcp_data_campaign", mockCampaignEvent);
            emitMockEvent("mcp_data_campaign", {...mockCampaignEvent, target: "remove_one"});
            emitMockEvent("mcp_data_campaign", {...mockCampaignEvent, target: "remove_all"});
        });

        act(() => {
            result.current.clearCampaigns();
        });
        
        expect(result.current.ready).toBeFalsy();
        expect(result.current.campaigns).toMatchSnapshot();
    });

    test("custom prioritization handler is called if passed.", () => {
        const handler = jest.fn().mockReturnValue([mockCampaignEvent]);
        const {result} = renderHook(() => useDataCampaign<TestCampaignType>(["my_target"], {displayLimit: 1, prioritizationAttributesInOrderOfImportance: ["text"]}, handler));

        act( () => {
            emitMockEvent("mcp_data_campaign", mockCampaignEvent);
        });
        
        expect(handler).toHaveBeenCalledTimes(1);
    });

});