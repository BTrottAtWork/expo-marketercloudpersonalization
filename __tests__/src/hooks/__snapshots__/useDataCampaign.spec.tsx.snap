// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDataCampaign Campaign fired for another target should render null 1`] = `[]`;

exports[`useDataCampaign No campaigns fired should render null 1`] = `[]`;

exports[`useDataCampaign continue to render hello world when single campaign is removed. 1`] = `
[
  {
    "campaignId": "my_id",
    "data": {
      "text": "hello world",
    },
    "target": "my_target",
  },
]
`;

exports[`useDataCampaign render null when all campaigns are removed. 1`] = `[]`;
